platform-hub-api/app/models/support_request_template.rb

```diff
@@ -0,0 +1,107 @@
+class SupportRequestTemplate < ApplicationRecord
+
+  include FriendlyId
+  include Audited
+
+  audited descriptor_field: :shortname
+
+  friendly_id :shortname, :use => :slugged
+
+  validates :shortname,
+    presence: true,
+    uniqueness: { case_sensitive: false }
+
+  validates :git_hub_repo,
+    presence: true
+
+  validates :title,
+    presence: true
+
+  validates :description,
+    presence: true
+
+  validates :form_spec,
+    presence: true
+
+  validates :git_hub_issue_spec,
+    presence: true
+
+  validate :check_schemas
+  validate :check_form_spec
+
+  class << self
+    attr_accessor :schemas, :form_field_types
+  end
+
+  @form_field_types = Set.new([
+    'text',
+    'textarea',
+    'number',
+    'checkbox',
+    'select',
+    'datetime-local',
+    'email',
+    'url'
+  ]).freeze
+
+  # TODO: some of the fields below need conditional checks for presence (e.g.
+  # the `options` field within the `fields` list should only be present if
+  # `field_type` is 'select').
+  # A request has been made to the ClassyHash library to support this:
+  # https://github.com/deseretbook/classy_hash/issues/22
+  @schemas = {
+    form_spec: {
+      'help_text' => String,
+      'fields' => [[
+        {
+          'id' => /\w+/i,
+          'label' => String,
+          'field_type' => SupportRequestTemplate.form_field_types,
+          'required' => TrueClass,
+          'placeholder' => [:optional, String],
+          'help_text' => [:optional, String],
+          'options' => [:optional, String],
+          'multiple' => [:optional, TrueClass]
+        }
+      ]]
+    },
+    git_hub_issue_spec: {
+      'title_text' => String,
+      'body_text_preamble' => [:optional, String]
+    }
+  }.freeze
+
+  private
+
+  def check_schemas
+    perform_hash_validation :form_spec, self.form_spec, SupportRequestTemplate.schemas[:form_spec]
+    perform_hash_validation :git_hub_issue_spec, self.git_hub_issue_spec, SupportRequestTemplate.schemas[:git_hub_issue_spec]
+  end
+
+  def perform_hash_validation field_name, hash, schema
+    hash_errors = []
+    success = ClassyHash.validate(hash, schema, errors: hash_errors, raise_errors: false, full: true)
+    unless success
+      self.errors.add(field_name, hash_validation_errors_to_string(hash_errors))
+    end
+  end
+
+  def hash_validation_errors_to_string errors
+    errors.map do |e|
+      "-- #{e[:full_path]} should be #{e[:message]}"
+    end.join(' ')
+  end
+
+  def check_form_spec
+    if self.form_spec
+      fields = self.form_spec['fields'] || self.form_spec[:fields]
```

Comment:
Good point. Actually, I can just use string keys version all the time. I originally added this because sometimes the check is on new data, and sometimes on data from the db. New data will (usually) be set from controller params (which is an indifferent access hash anyway), but data from the db will always be string keys (very annoying really, but a quirk of the postgres adapter).

Will change to use string keys only.